airbyte-integrations/bases/base-normalization/normalization/transform_catalog/transform.py

```diff
@@ -139,6 +101,269 @@ def extract_schema(profiles_yml: dict) -> str:
+        sql += "{},\n  ".format(parent_hash_id)
+    for field in properties.keys():
+        sql += quote(field, integration_type) + ",\n  "
+    sql += "_airbyte_emitted_at,\n  "
+    sql += hash_id
+    sql += "\nfrom numbered where row_num = 1\n"
+    if len(path) > 1:
+        sql += "-- {} from {}\n  ".format(name, "/".join(path))
+    output_sql_table(output, schema, sql_file_name, sql, path)
+
+    # Generate children models
+    previous_sql_file_name = sql_file_name
+    table = "{{ ref('" + previous_sql_file_name + "') }}"
+    for field in properties.keys():
+        if is_airbyte_column(field):
+            pass
+        elif is_combining_node(properties[field]):
+            # TODO: merge properties of all combinations
+            pass
+        elif "type" not in properties[field] or is_object(properties[field]["type"]):
+            process_nested_property(
+                output=output,
+                integration_type=integration_type,
+                path=path,
+                json_col=f"'{field}'",
+                tables_in_schema=tables_in_schema,
+                properties=properties[field],
+                raw_schema=raw_schema,
+                schema=schema,
+                table=table,
+                parent_hash_id=hash_id,
+                inject_sql=f"where {field} is not null\n",
+                field=field,
+            )
+        elif is_array(properties[field]["type"]) and "items" in properties[field]:
+            quoted_field = quote(field, integration_type, in_jinja=True)
+            process_nested_property(
+                output=output,
+                integration_type=integration_type,
+                path=(path),
+                json_col="unnested_column_value('_airbyte_data')",
+                tables_in_schema=tables_in_schema,
+                properties=properties[field]["items"],
+                raw_schema=raw_schema,
+                schema=schema,
+                table=table,
+                parent_hash_id=hash_id,
+                inject_sql="\ncross join {} as _airbyte_data\nwhere {} is not null\n".format(
+                    jinja_call(f"unnest({quoted_field})"), quote(field, integration_type)
+                ),
+                field=field,
+            )
+
+
+def process_nested_property(
+    output: str,
+    integration_type: str,
+    path: list,
+    json_col: str,
+    tables_in_schema: Dict[str, Set[str]],
+    properties,
+    raw_schema,
+    schema,
+    table,
+    parent_hash_id: str,
+    inject_sql: str,
+    field: str,
+):
+    children = find_properties_object(path=[], field=field, properties=properties, integration_type=integration_type)
+    for child in children:
+        child_name = child
+        if child_name in tables_in_schema[schema]:
+            if integration_type == "postgres" and len(child) >= 59:
+                child_str = child[0:55]
+            else:
+                child_str = child
+            for i in range(1, 100):
```

Comment:
what is this doing? it's handling naming collisions of child tables, i think? i think i believe that there's a real problem here, but can you help me understand exactly what it is and how it is being solved?

Reply:
yes, it's handling name collisions from the currently processed catalog...

In Stripe connector for example, in the `charges` stream, there is the concept of `card` multiple times and depending on where it's from, it has different schemas:
- charges/card
- charges/source/card
- charges/payment_method_details/card

![Screenshot 2021-02-12 at 11 15 21](https://user-images.githubusercontent.com/7718175/107755856-a685c680-6d23-11eb-947e-b6cc73b552f5.png)

The "exploded" tables are:
- named after the nested column name (in my example, it's `card`)
- If I name them after the "path" to the nested column, it's making super long names that easily overflows the number of character limits allowed
- if the name was already used by this stream or another stream in the same catalog, then it just adds an integer `i` counting allowing up to 100 collisions (top-level stream names have priority over nested children) 

Sometimes, the collisions have the same schema (below with `address`) and could be unioned/merged into a single table but I am leaving it to the user to do that extra step afterward... So if someone wants to merge the different `card` tables into a single one and keep only a subset of common fields or filled in with blanks, then it's up to them...

![Screenshot 2021-02-12 at 11 13 16](https://user-images.githubusercontent.com/7718175/107756224-3592de80-6d24-11eb-9f25-c44d469c35fb.png)

Reply:
discussed with chris he will create an issue for this. we can live with this for now but need to fix quickly

Reply:
Issue created here: https://github.com/airbytehq/airbyte/issues/2055